flybywiresim/a32nx
# refactor(isis): Convert ISIS to React

Fixes #5222
Contributes to: #1233

## Summary of Changes

The ISIS is now coded using React. All of the existing functionality should be implemented more or less identically with some small additions including the Bugs page.

Note: At the moment, the ISIS is still using data from the common ADIRS, this is to be changed.

Implemented features:
- [x] Attitude indicator
  - [x] Pitch scale (translation)
  - [x] Excessive pitch arrows
  - [x] Roll scale
  - [x] Round off corners
- [x] Aircraft symbol
- [x] Airspeed indicator
- [x] Altimeter
  - [x] Static
  - [x] Scale
  - [x] Show NEG if negative
- [x] Pressure
- [x] Mach number
- [x] Metric altitude indicator (optional configuration)
  - [x] Possibly serialize to local variables
- [x] LS
- [x] ATT RST
- [x] Startup
- [x] Ensure electrical power logic is correct
- [x] Brightness
- [x] Clean up CSS
- [x] Align everything perfectly
- [ ] Use data from correct ADIRU

Stuff not covered by this PR: 
- Indications for loss of altitude, airspeed, attitude, localizer, glideslope, etc. through flags is not implemented
- Excessive movement during the alignment phase does not trigger ATT:RST flag

## Screenshots (if necessary)



## References

- FCOM
- For bugs: https://www.youtube.com/watch?v=U1bCkxL1QmY
- General cockpit view: https://ccntservice.airbus.com/apps/cockpits/a320/
- For ATT RST (A330/A340): https://www.youtube.com/watch?v=o9yh7WFDcBQ&t=479s
- For NEG text: https://youtu.be/cv41hm4JEB8?t=213




## Additional context



Discord username (if different from GitHub): BlueberryKing#6641

## Testing instructions


- In general: The functionality should match the functionality that is currently implemented.
- Artificial horizon
  - Ensure pitch attitude is correctly represented. Compare to PFD.
  - Try bringing the aircraft into an excessive attitude (either very high pitch or very low pitch) and check for arrows showing the direction to reduce the pitch attitude.
  - Ensure the aircraft's roll state is correctly represented on the roll axis
- Speed indication
  - Ensure indicated speed matches the one shown on the PFD
- Altitude indication
  - Ensure the altitude tape shows the altitude shown on the PFD
  - Ensure the digital altitude readout matches the aircraft's current altitude
  - Roughly ensure altitude in ft converts to the altitude shown in meters.
- Mach indicator
  - The Mach indicator on the bottom left of the instrument should become visible above Mach 0.5 and disappear again once the speed drops below Mach 0.45
- Pressure indicator
  - Ensure the pressure value in hPa is adjustable using the baro knob on the bottom right. The inHg indication should adjust accordingly.
  - By pressing the baro knob, the pressure indication toggle between "STD" and hPa/inHg.
- Bugs
  - Set a speed bug on the bugs page, make sure to enable it by pressing the baro knob (the OFF text should disappear), and check whether it shows up as a cyan bar on the speed tape.
  - Similarly, set an altitude bug, ensure it shows up as a cyan bar on the altitude tape. Different from the speed tape, if an altitude bug is set to a value that is indicated by a number, i.e every 500 ft, this number should be in a cyan box. When the aircraft's current altitude is within 100 ft of a bug value, the digital altitude readout box should turn cyan.
  - After 15s of no pilot input, the bugs page should automatically disappear and the artificial horizon should be shown again.
- LS
  - Fly an ILS, enable the LS scales by pressing the "LS" button above the display unit. Ensure the localizer and glideslope diamond positions represent the aircraft's position relative to the respective beams. It is probably easiest to compare them to the PFD.
- ATT RST
  - Press and hold the "ATT" button below the display unit for 2s, then let go. The artificial horizon should disappear in favor of an "ATT XXs" seconds message counting down from 10s. After the 10s have passed, the artificial horizon should return.


## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page
